File: hospital-backend/src/main/java/com/hospital/dao/UserDAO.java
package com.hospital.dao;

import com.hospital.model.User;
import java.util.List;

public interface UserDAO {
    List<User> getAll();
    User  getById(int id);
    User  getByEmail(String email);       // used by LoginServlet for credential lookup
    User  getByUsername(String username);
    List<User> getByRole(String roleName);// used by admin to list all doctors/patients/staff
    int   insert(User user);
    int   update(User user);
    int   delete(int id);
    int   setActive(int userId, boolean active);          // activate / deactivate
    int   updateLastLogin(int userId);                    // called on successful login
}

File: hospital/model/dao/UserDAO.js
/**
 * DAO Interface: UserDAO
 * Core auth table — shared across all roles.
 */
export default class UserDAO {
  async getAll()                     { throw new Error('UserDAO.getAll() not implemented'); }
  async getById(id)                  { throw new Error('UserDAO.getById() not implemented'); }
  async getByEmail(email)            { throw new Error('UserDAO.getByEmail() not implemented'); }
  async getByUsername(username)      { throw new Error('UserDAO.getByUsername() not implemented'); }
  async getByRole(roleId)            { throw new Error('UserDAO.getByRole() not implemented'); }
  async create(user)                 { throw new Error('UserDAO.create() not implemented'); }
  async update(id, data)             { throw new Error('UserDAO.update() not implemented'); }
  async delete(id)                   { throw new Error('UserDAO.delete() not implemented'); }
  async setActive(id, isActive)      { throw new Error('UserDAO.setActive() not implemented'); }
  async updateLastLogin(id)          { throw new Error('UserDAO.updateLastLogin() not implemented'); }
  async verifyPassword(id, hash)     { throw new Error('UserDAO.verifyPassword() not implemented'); }
}
